Upstream Telephone Bearer Service (TBS) server and client

Thalley opened the pull request:
Upstreams the Telephone Bearer Service server and client from the topic-le-audio branch.

This is mainly a copy of existing and already reviewed code, but updates to pass CI and where CCP has been replace with TBS_CLIENT, as that is more descriptive of the actual implementation, as well as being more in line with some of the other LE Audio service names. 

fixes https://github.com/zephyrproject-rtos/zephyr/issues/41594

Thalley committed:
Bluetooth: Audio: Add Telephone Bearer Service server

Add the telephone bearer service server implementation.
This support multiple service instances as well as the
generic telephone bearer service, but is still a work in
progress and should be treated as such.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add TBS shell module

Add a shell module for the TBS.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add Telephone Bearer Service client

Add a client for the telephone bearer service (TBS).

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add Telephone Bearer Client shell module

Add a shell module for the TBS client.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add TBS server and client BSIM tests

Add babblesim tests for the TBS server and client.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

jhedberg commented on subsys/bluetooth/audio/Kconfig.tbs:
```diff
@@ -0,0 +1,213 @@
+
+config BT_TBS
+	bool "Telephone Bearer Service Support"
+	default n
```
Minor: n is the default default, so this line is unnecessary

jhedberg commented on subsys/bluetooth/audio/tbs_client.c:
```diff
@@ -0,0 +1,2166 @@
+		if (inst->calls[i].index == BT_TBS_FREE_CALL_INDEX) {
+			return true;
+		}
+	}
+	return false;
```
Minor: add an empty line before the return

jhedberg commented on subsys/bluetooth/audio/tbs_client.c:
```diff
@@ -0,0 +1,2166 @@
+{
+	uint8_t item_len;
+	uint8_t uri_len;
+	const size_t mimimum_item_len = sizeof(call->call_info) + BT_TBS_MIN_URI_LEN;
```
Seems like a misspelling of "minimum". Also, I'd just call it `min_item_len`. Also, move it as the first variable definition instead of having a long line in the middle of the definitions.

jhedberg commented on subsys/bluetooth/audio/tbs_client.c:
```diff
@@ -0,0 +1,2166 @@
+			      const void *data, uint16_t length)
+{
+	uint16_t handle = params->value_handle;
+	struct bt_tbs_instance *tbs_inst = lookup_inst_by_handle(conn,
+								       handle);
```
Unnecessary line split?

jhedberg commented on subsys/bluetooth/audio/tbs_client.c:
```diff
@@ -0,0 +1,2166 @@
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static int tbs_client_common_call_control(struct bt_conn *conn, uint8_t inst_index,
+				   uint8_t call_index, uint8_t opcode)
```
The split line alignment doesn't look right

Thalley replied:
It wasn't :)

jhedberg commented on subsys/bluetooth/audio/tbs_client.c:
```diff
@@ -0,0 +1,2166 @@
+					      &common, sizeof(common), false);
+}
+
+static uint8_t read_bearer_provider_name_cb(struct bt_conn *conn,
+						uint8_t err,
+						struct bt_gatt_read_params *params,
+						const void *data,
+						uint16_t length)
+{
```
Misalignment with the split lines here as well (there are more places in this file, but I won't comment on them separately)

jhedberg commented on subsys/bluetooth/audio/tbs_client.c:
```diff
@@ -0,0 +1,2166 @@
+				parse_string_value(data, length,
+					CONFIG_BT_TBS_MAX_PROVIDER_NAME_LENGTH);
+			BT_DBG("%s", log_strdup(provider_name));
+		}
+		inst->busy = false;
```
minor: add an empty line after {

jhedberg commented on subsys/bluetooth/audio/tbs_client.c:
```diff
@@ -0,0 +1,2166 @@
+				BT_DBG("Invalid length");
+				cb_err = BT_ATT_ERR_INVALID_ATTRIBUTE_LEN;
+			}
+		}
+		inst->busy = false;
```
Same here. Won't comment in other places if I see it again.

Thalley replied:
Fixed it everywhere. 

This is one of the rules that I referenced in https://github.com/zephyrproject-rtos/zephyr/pull/40707#discussion_r761788085 that are unwritten and is only enforced by a few people. 

If we truly want to enforce such specific rules for the code base, those should be checked for in CI, or at least be written down in the official coding guidelines.

jhedberg replied:
I suspect it's going to be seen as a too minor thing to be enforced tree-wide.

jhedberg replied:
I also think that while tree-wide consistency is important, the importance of consistency is inversely proportional to the size of the scope that's being viewed (e.g. it's more important for code within a single c-file to be consistent than two c-files from different subsystems in the same tree).

Thalley commented on subsys/bluetooth/audio/Kconfig.tbs:
```diff
@@ -0,0 +1,212 @@
+	help
+	  This option enables support for Generic Telephone Bearer Service.
+
+config BT_TBS_PROVIDER_NAME
+	string "Telephone Bearer Service Provider Name"
+	default "Unknown"
+	help
+	  Sets the name of the service provider for the bearer.
+
+config BT_TBS_UCI
+	string "Telephone Bearer Service Uniform Caller Identifier (UCI)"
+	default "un000"
+	help
+	  Sets the UCI of the bearer. See
+		https://www.bluetooth.com/specifications/assigned-numbers/uniform-caller-identifiers/
+		for a table of valid UCIs.
+
+config BT_TBS_TECHNOLOGY
+	int "Telephone Bearer Service Technology"
+	range 1 10
+	help
+	  Sets the technology used for the bearer, e.g. GSM, LTE and 5G.
+	  1 : 3G
+	  2 : 4G
+	  3 : LTE
+	  4 : Wi-Fi
+	  5 : 5G
+	  6 : GSM
+	  7 : CDMA
+	  8 : 2G
+	  9 : WCDMA
+	  10: IP
```
Agreed. The idea is definitely also to put these into runtime registrations as we did for e.g. the BT_VCS implementation. However, that will need to be fixed in a later PR, as the idea of this, as with any of the other LE Audio services, is that we simply upstream what we have, so that it is publicly available.

tejlmand replied:
Would a choice be clearer to users so it's obvious what they are selecting / what is selected ?

In `menuconfig`:
```
Telephone Bearer Service Technology (Wi-Fi)
```
and:
```
( ) 3G
( ) 4G
( ) LTE
(X) Wi-Fi
```

and then in `prj.conf`:
```
CONFIG_BT_TBS_TECHNOLOGY_WIFI=y
```
vs:
```
CONFIG_BT_TBS_TECHNOLOGY=4 # <-- What does this mean, damn I need to look at the docs :(
```

jhedberg replied:
@tejlmand yeah, that sounds like a better approach, at least as long as this comes from Kconfig

Thalley replied:
I'd rather not start modifying things that aren't broken when simply upstreaming an existing implementation. The next step for this would be to dynamically register the TBS instances, which would solve this.

Thalley commented on subsys/bluetooth/audio/Kconfig.tbs:
```diff
@@ -0,0 +1,212 @@
+	help
+	  Sets the maximum number of calls the service supports per bearer.
+
+config BT_TBS_BEARER_COUNT
+	int "How many bearer instances the device instantiates"
```
That is indeed a missing feature. Will need to be fixed later, likely as part of a PR that allows dynamic registration of TBS instances.

Thalley commented on subsys/bluetooth/audio/Kconfig.tbs:
```diff
@@ -0,0 +1,212 @@
+	  Sets the maximum length of the call URI supported. Note that if this
+	  value is lower than a call URI, the call request will be rejected.
+
+config BT_TBS_MAX_PROVIDER_NAME_LENGTH
+	int "The maximum length of the bearer provider name"
```
The provider name is not static. One is setting the initial value, and the other sets the size of the array that holds the name.

Thalley commented on subsys/bluetooth/audio/tbs.c:
```diff
@@ -0,0 +1,2485 @@
+				if (call->index == BT_TBS_FREE_CALL_INDEX) {
+					continue;
+				}
+				(void)net_buf_simple_add_u8(&read_buf,
+							    call->index);
+				(void)net_buf_simple_add_u8(&read_buf,
+							    call->state);
+				(void)net_buf_simple_add_u8(&read_buf,
+							    call->flags);
```
Honestly at this point I don't even know anymore :D 

Some reviewers require that all return values are either explicitly ignored by (void), or otherwise stored in variables, and some say it is okay to leave some out. 

There is a MISRA rule that states that if the function returns error information, that shall be tested (which we also do not do in many places). That is not the case here though. 

Perhaps we should conclude on whether we always shall explicitly handle all return values (e.g. by adding `(void)`, or whether it is okay to not do it), as that is clearly something that reviewers have different opinions on (we also have a bunch of other unwritten rules that are "enforced" by different reviewers in terms of coding style...).

jhedberg replied:
The main thing I care about is subsystem-wide consistency. Either this should be done everywhere, or nowhere.

jhedberg replied:
We also have the quite useful tool `__must_check` that's currently nowhere nearly sufficiently used in the tree. If it was, I think the importance of these void casts would get reduced (for me they at least have the effect of reduced readability, i.e. they cause me to stop for a sec and wonder "what's going on here").

jhedberg replied:
And just to follow up on that: using `__must_check` implies that it's almost always a bug to ignore the return value, and a void cast is therefore mandated to make the ignoring explicit and clearly intentional. In the case of these net_buf APIs the return value is a purely optional convenience that can be used if it makes sense, but in the majority of uses it's not really needed or useful, kind of like an optional input parameter where you can pass NULL (or similar) if you don't care. Therefore forcing the majority of users of the API to do explicit void casts doesn't feel right to me.

I understand where MISRA is coming from with this, but the core problem is that standard C doesn't have a way of declaring optional vs mandatory return values, so the safe bet is to treat everything as mandatory. However, we can do better than that with the help of `__must_check`, and if it was properly used then the forced casts would only happen for APIs where the return value has been annotated as mandatory.

jhedberg replied:
I meant per function. If `(void)` is required for a function to ignore its return value in one place, then it should be used everywhere, and vice-versa. Also, I don't think MISRA cares about the meaning of the return value (if it's an error code or something else). As long as there is a return value you must to do something with it according to MISRA (a void cast being one option of "doing something").

> __must_check is a great thing, but should only be used for functions that may return an error value

Why do you say that? I don't think that's accurate. E.g. `net_buf_ref()` doesn't return an error value but uses `__must_check` in order to make sure that there are no reference leaks in the system. Likewise, it'd make sense to apply `__must_check` on any allocator-style functions or functions like `k_fifo_get()` - not just to catch allocation errors, but to make sure that the returned value/memory isn't leaked.

Anyway, this is getting a bit beside the point since the `net_buf_simple_add_*()` APIs don't fall into those categories. As I said, the purpose of the return value there is to be a friendly optional feature that may be used in some rare situations, and forcing the use of `(void)` on the majority of users who don't need this feature doesn't feel quite right.

Thalley replied:
> Also, I don't think MISRA cares about the meaning of the return value (if it's an error code or something else)

The MISRA rule I'm thinking of, is this (from https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#main-rules)

![image](https://user-images.githubusercontent.com/2520413/145200942-a2e94330-3d2b-4e02-b68e-07e6b98c10ea.png)

> Why do you say that? I don't think that's accurate. E.g. net_buf_ref() doesn't return an error value but uses __must_check in order to make sure that there are no reference leaks in the system.

That's true, it is also very useful for such examples. 

> Anyway, this is getting a bit beside the point since the net_buf_simple_add_*() APIs don't fall into those categories. 

Personally I don't see any reason why we wouldn't do it for `net_buf_simple_add_*` if we do it for e.g. `memset`. I also don't think it harms anything that the code explicitly  ignore the return value, but if you really want it gone, I'll remove it.

jhedberg replied:
It might be just my personal way of thinking about explicit casts. But whenever I see them while reading code it makes me stop and think "what's going on here? Is this right?". That's quite wasteful since it's almost always fine to ignore these return values. Anyway, wrt this specific case, there are three acceptable options from my perspective, in order of preference:

1. Don't do the explicit cast.
2. If there are no users of the return value in the tree just make these APIs return void.
3. Do the explicit cast everywhere for these functions.

This PR as it stands now doesn't do any of those options.

jhedberg replied:
Regarding that MISRA rule, I'm then confused why explicit casts have been introduced in Zephyr for libc functions that return a value but it's not an error (like memset or memcpy). I wonder if there's some other rule which requires it..

Thalley replied:
> Regarding that MISRA rule, I'm then confused why explicit casts have been introduced in Zephyr for libc functions that return a value but it's not an error (like memset or memcpy). I wonder if there's some other rule which requires it..

Agreed. I figured that someone some time added that requirement, or at least recommendation (we still have some uses of those which aren't explicitly cast), but they do not make any more sense than explicit casts for most of the net_buf functions. 

> This PR as it stands now doesn't follow any of those options. 

I'll remove it for now. Then, as part of the discussion for more general code guidelines/rules for the BT subsystem, we can revisit it.

asbjornsabo commented on tests/bluetooth/bsim_bt/bsim_test_audio/src/tbs_client_test.c:
```diff
@@ -0,0 +1,232 @@
+
+static struct bt_tbs_client_cb tbs_client_cbs = {
+	.discover = tbs_client_discover_cb,
+	.originate_call = NULL,
```
It seems for both client and server that there is more that could be covered by tests.

Thalley replied:
Absolutely, which is also why we have an pending task for that :)

Thalley commented on include/bluetooth/audio/tbs.h:
```diff
@@ -0,0 +1,737 @@
+ * @return true if the call request was accepted and remote party is alerted.
+ */
+typedef bool (*bt_tbs_originate_call_cb)(struct bt_conn *conn,
+					 uint8_t call_index,
```
Github, yes :)

Thalley commented on include/bluetooth/audio/tbs.h:
```diff
@@ -0,0 +1,737 @@
+
+/**
+ * @brief Callback function for client originating a call.
+ *
```
Agreed. 

I wonder if the word "originating" is the issue, as I don't think it is as confusing for e.g. the `bt_tbs_terminate_call_cb` that uses the same phrasing.

Thalley commented on include/bluetooth/uuid.h:
```diff
@@ -1781,6 +1862,78 @@ struct bt_uuid_128 {
  */
 #define BT_UUID_CCID \
 	BT_UUID_DECLARE_16(BT_UUID_CCID_VAL)
```
Personally I don't think we need to change it. It may be independent of TBS, but it is placed in the same order as the full official list of UUIDs. 

Personally I would have defined it separately, instead of where it is now, but I think it is better to simply follow the SIG ordering.

Thalley commented on subsys/bluetooth/audio/Kconfig.tbs:
```diff
@@ -0,0 +1,212 @@
+
+config BT_GTBS
+	bool "Generic Telephone Bearer Service Support"
+	default y
```
Nope. I'll add a TODO (fixing it would require a significant amount of work).

asbjornsabo reviewed (dismissed):
There is a number of things here, large and small, that could be improved.  (I guess that is just proof that we have learnt and improved over the years since this code was started.)
Some of these things there is already plans for, like the API update.
But the code is fundamentally sound (IMO) and working as is.

Since the purpose of this PR is to get the code into the main branch (it is a wish from others to get rid of the topic-le-audio branch) I think we should just merge it, and do the improvements afterwards.  (That will also give us a much cleaner history, with no changes "hidden" in the transition from the topic-le-audio branch to the main branch.)

Thalley commented on subsys/bluetooth/audio/tbs_internal.h:
```diff
@@ -0,0 +1,278 @@
+#define BT_TBS_CALL_OPCODE_ORIGINATE               0x04
+#define BT_TBS_CALL_OPCODE_JOIN                    0x05
+
+/* Remote Control Points */
```
Added some text

Thalley commented on subsys/bluetooth/audio/tbs_internal.h:
```diff
@@ -0,0 +1,278 @@
+struct bt_tbs_call {
+	uint8_t index;
+	uint8_t state;
+	uint8_t flags;
```
It could. Would require a significant update in the code, and would rather leave that for a future update.

asbjornsabo replied:
OK

Thalley commented on include/bluetooth/audio/tbs.h:
```diff
@@ -0,0 +1,737 @@
+ * @return int          BT_TBS_RESULT_CODE_* if positive or 0,
+ *                      errno value if negative.
+ */
+int bt_tbs_remote_answer(uint8_t call_index);
```
That is true. 

I don't think particular one can ever be handled via a callback, as the time from starting/placing/originating a call and to a call is actually picked up (or terminated) may be a long time.

Thalley commented on subsys/bluetooth/shell/tbs_client.c:
```diff
@@ -0,0 +1,785 @@
+#include <bluetooth/audio/tbs.h>
+
+#include "bt.h"
+
```
Correct. 

The shell module is far from "good", and requires that debugging is enabled before it is actually useful (as the debug log will provide the information necessary). 

Will require a significant update.

asbjornsabo replied:
Make a note of it for later ...

Thalley commented on subsys/bluetooth/audio/tbs.c:
```diff
@@ -0,0 +1,2485 @@
+	const struct bt_gatt_service_static *service_p;
+};
+
+#if IS_ENABLED(CONFIG_BT_GTBS)
```
See https://github.com/zephyrproject-rtos/zephyr/pull/40707#discussion_r761808581

Thalley commented on subsys/bluetooth/audio/tbs.c:
```diff
@@ -0,0 +1,2485 @@
+NET_BUF_SIMPLE_DEFINE_STATIC(read_buf, READ_BUF_SIZE);
+
+static struct tbs_service_inst svc_insts[CONFIG_BT_TBS_BEARER_COUNT];
+static struct gtbs_service_inst gtbs_inst;
```
I guess the guard is missing, yes :) 
Will require significant amount of work to fix.

Thalley replied:
https://github.com/zephyrproject-rtos/zephyr/issues/42487

Thalley commented on subsys/bluetooth/audio/tbs.c:
```diff
@@ -0,0 +1,2485 @@
+		return BT_TBS_RESULT_CODE_INVALID_CALL_INDEX;
+	}
+
+	call->index = BT_TBS_FREE_CALL_INDEX;
```
Yes, that is how the state machine is defined.

MariuszSkamra commented on include/bluetooth/audio/tbs.h:
```diff
@@ -328,4 +328,411 @@ void bt_tbs_register_cb(struct bt_tbs_cb *cbs);
+
+struct bt_tbs_client_cb {
+	bt_tbs_client_discover_cb            discover;
+	bt_tbs_client_cp_cb                  originate_call;
+	bt_tbs_client_cp_cb                  terminate_call;
+	bt_tbs_client_cp_cb                  hold_call;
+	bt_tbs_client_cp_cb                  accept_call;
+	bt_tbs_client_cp_cb                  retrieve_call;
+	bt_tbs_client_cp_cb                  join_calls;
+
+	bt_tbs_client_read_string_cb         bearer_provider_name;
+	bt_tbs_client_read_string_cb         bearer_uci;
+	bt_tbs_client_read_value_cb          technology;
+	bt_tbs_client_read_string_cb         uri_list;
+	bt_tbs_client_read_value_cb          signal_strength;
+	bt_tbs_client_read_value_cb          signal_interval;
+	bt_tbs_client_current_calls_cb       current_calls;
+	bt_tbs_client_read_value_cb          ccid;
+	bt_tbs_client_read_value_cb          status_flags;
+	bt_tbs_client_read_string_cb         call_uri;
+	bt_tbs_client_call_states_cb         call_state;
+	bt_tbs_client_read_value_cb          optional_opcodes;
+	bt_tbs_client_termination_reason_cb  termination_reason;
+	bt_tbs_client_read_string_cb         remote_uri;
+	bt_tbs_client_read_string_cb         friendly_name;
```
Does it make sense to make the client features configurable via Kconfig to reduce the memory consumption?
Simplest and sufficient for most of the users configuration would be to Accept and Terminate and track Call State.

e.g most of the TBS functionality is not used by implementation that has no display.

Thalley replied:
Absolutely! 
We actually even already have https://github.com/zephyrproject-rtos/zephyr-bt-sig/issues/29 and https://github.com/zephyrproject-rtos/zephyr-bt-sig/issues/28 for such ideas. 

We definitely need to be able to disable anything not mandatory.

MariuszSkamra replied:
Great!

MariuszSkamra commented on include/bluetooth/audio/tbs.h:
```diff
@@ -328,4 +328,411 @@ void bt_tbs_register_cb(struct bt_tbs_cb *cbs);
+ */
+int bt_tbs_client_read_friendly_name(struct bt_conn *conn, uint8_t inst_index);
+
+/** @brief Read the supported opcode of a TBS instance.
```
Shouldn't this be handled internally, and not exposed by API? My feeling is that if client supports (see my other comment related to client configurable features) any of the *optional* CP operations the characteristic could be read internally. So that if API user will try to e.g. Local Hold the implementation would respond with an error immediately (not even writing to the CP).

Thalley replied:
That's not a bad idea. The API was implemented to be as flexible as possible, and not do anything unnecessary and following the spec defined procedures. e.g. if you are doing a client that only uses the mandatory features, you wouldn't need to do this at all. 

Doing it autonomously would be somewhat conflicting with your other comment https://github.com/zephyrproject-rtos/zephyr/pull/40707#discussion_r764671350

MariuszSkamra replied:
It would not conflict. It could be done autonomously if the Client would compile the code with Hold and/or Join features enabled.

Thalley replied:
Good point :) It's definitely something we can change in the future. This does not need to be an API indeed.

rymanluk replied:
how do we track changes we want to do in the future?

Thalley replied:
@rymanluk Excellent question. Let's discuss that in tomorrow meeting. We are tracking it internally in Nordic, but we may need to double track it somehow.

MariuszSkamra commented on include/bluetooth/audio/tbs.h:
```diff
@@ -328,4 +328,411 @@ void bt_tbs_register_cb(struct bt_tbs_cb *cbs);
+ *
+ * @return int          0 on success, GATT error value on fail.
+ */
+int bt_tbs_client_discover(struct bt_conn *conn, bool subscribe);
```
No such thing as a stupid question ^^ Why not subscribe by default? Most of the TBS characteristics have Notify property, thus once read *seamlessly* by client implementation right after connection could be notified if there's any change by the server.

This would reduce this API by removing all the read request functions like bt_tbs_client_read_bearer_provider_name.
If there's possiblity to notify I would stick to it as ATT Read operation is blocking while ATT Notification can be sent anytime.

Thalley replied:
This is some very old code where we wanted to give that flexibility. It is likely something we will remove, and always subscribe by default. Since this is just moving the current implementation, this won't be fixed in this PR, but will be fixed in a future update.

Thalley replied:
https://github.com/zephyrproject-rtos/zephyr/issues/42488

MariuszSkamra commented on include/bluetooth/audio/tbs.h:
```diff
@@ -0,0 +1,331 @@
+ * @return int          BT_TBS_RESULT_CODE_* if positive or 0,
+ *                      errno value if negative.
+ */
+int bt_tbs_accept(uint8_t call_index);
```
If I were telephone service implementer that would expose the functionality over this API that would be a difficult for me TBH. When I should call this function (and related)? I have mixed feelings because it looks like it shall be called to accept the call as the name suggests.

If that's Server side API why this function isn't called e.g. *bt_tbs_call_accepted* ?
I guess the intention is not to accept the call by calling this function but rather indicate that the call has been accepted by a service that implements telephone functionality.

Or API user should perform the call acceptance from accept_call callback context? I guess not, because it would be odd that Bluetooth defines API for managing telephone service

Thalley replied:
This is a server side API indeed. `bt_tbs_call_accepted` is a better name, and will update :) 

We don't want to use a callback for this, as the time between the server putting a call in the calling state, and when the call is actually accepted can be very long.

MariuszSkamra replied:
> This is a server side API indeed. `bt_tbs_call_accepted` is a better name, and could be updated :) However, if we update this, then `bt_tbs_hold`, `bt_tbs_retrieve` etc. should similar be updated. Should we rename all the server APIs for controlling the call state be of the pattern `bt_tbs_call_Xed` instead of bt `bt_tbs_X`?

I do not want to push, but that's my feeling.
The Server API user by calling those functions wants to notify the call state changed (e.g. was accepted) not to change the call state (e.g. to accept).

Just like:
bt_conn_disconnect <- initiate termination
bt_conn_disconnected <- already terminated, tell anyone
^^

> 
> We don't want to use a callback for this, as the time between the server putting a call in the calling state, and when the call is actually accepted can be very long.
Understood, I wanted to show this can be misused.

Thalley replied:
> The Server API user by calling those functions wants to notify the call state changed (e.g. was accepted) not to change the call state (e.g. to accept).

The calls and the states reside in the TBS instance at this moment, so these function do change the state and do not just notify the state changed.

MariuszSkamra replied:
But the API user does not know the TBS implementation internals, does not have to

Thalley replied:
> But the API user does not know the TBS implementation internals, does not have to

I'm sorry, I don't think I understand what you mean by this. Can you elaborate?

MariuszSkamra replied:
> The calls and the states reside in the TBS instance at this moment, so these function do change the state and do not just notify the state changed.

I mean the API user - e.g. telephone implementation will most likely hold the calls and it's states is it's own structures. API user does not have to know whether the TBS holds it's copy or not, it does not matter. So the function name shall indicate that API user should call this function when the call state changed - because this function will notify about this change. The way how it's implemented (it stores the copy of the state as you said) does not matter.

Thalley replied:
> I mean the API user - e.g. telephone implementation will most likely hold the calls and it's states is it's own structures. 

That is not how TBS is currently implemented, and based on the discussions in the F2F it also haven't been decided whether that should be the case or not yet.

alwa-nordic replied:
I think it's natural that methods have names in verb-form, and "bt_tbs_call_accepted" can be parsed in verb-form as "call the accepted thing", which is wrong. Whatever we go for, we should minimize ambiguity. I prefer "bt_tbs_accept" or "bt_tbs_set_call_accepted".

jhedberg commented on subsys/bluetooth/audio/tbs.c:
```diff
@@ -0,0 +1,2481 @@
+	bool notify_current_calls;
+	bool notify_call_states;
+	bool pending_signal_strength_notification;
+	struct k_work_delayable reporting_interval_work;
+
+	/* TODO: The TBS (service) and the Telephone Bearers should be separated
+	 * into two different instances. This is due to the addition of GTBS,
+	 * where we now are in a state where this isn't a 1-to-1 correlation
+	 * between TBS and the Telephone Bearers
+	 */
+	struct bt_gatt_service *service_p;
+};
+
+struct gtbs_service_inst {
+	/* Attribute values */
+	char provider_name[CONFIG_BT_TBS_MAX_PROVIDER_NAME_LENGTH];
+	char uci[BT_TBS_MAX_UCI_SIZE];
+	uint8_t technology;
+	uint8_t signal_strength;
+	uint8_t signal_strength_interval;
+	uint8_t ccid;
+	uint16_t optional_opcodes;
+	uint16_t status_flags;
+	struct bt_tbs_in_uri incoming_uri;
+	struct bt_tbs_in_uri friendly_name;
+	struct bt_tbs_in_uri in_call;
+
+	/* Instance values */
+	bool notify_current_calls;
+	bool notify_call_states;
+	bool pending_signal_strength_notification;
+	struct k_work_delayable reporting_interval_work;
+
+	/* TODO: The TBS (service) and the Telephone Bearers should be separated
+	 * into two different instances. This is due to the addition of GTBS,
+	 * where we now are in a state where this isn't a 1-to-1 correlation
+	 * between TBS and the Telephone Bearers
+	 */
+	const struct bt_gatt_service_static *service_p;
+};
```
It seems to me like it might make sense to make the GTBS struct a parent struct of the TBS struct, i.e. so that you have a GTBS struct as the first element of the TBS struct and the TBS struct that way extends the GTBS struct. This would help eliminate the void pointer stuff you have with casting between these types (you'd always pass a GTBS pointer and then when TBS is needed you use CONTAINER_OF() to get it).

Thalley replied:
Yes, there certainly are a lot of common fields between the two. 

I tried doing a common struct before, but the number of differences between the two made me split it into two, but maybe I should revisit the idea. 
List of differences:

- uri_scheme_list only in TBS
- terminate_reason only in TBS
- index only in TBS
- calls only in TBS

So it would definitely make sense to have a base struct for GTBS and have the TBS struct use that. 

I'll add a note at the struct for this.

tejlmand commented on subsys/bluetooth/audio/Kconfig.tbs:
```diff
@@ -0,0 +1,155 @@
+	  Sets the number of TBS instances that are instantiated
+
+config BT_TBS_SERVICE_COUNT
+	int "How many bearer service instances the device instantiates"
```
Maybe:
```suggestion
	int "Number of instantiated bearer service instances"
```

tejlmand commented on subsys/bluetooth/audio/Kconfig.tbs:
```diff
@@ -0,0 +1,155 @@
+	  then the maximum length should be maximum 512 / BT_TBS_BEARER_COUNT.
+
+config BT_TBS_AUTHORIZATION
+	bool "Enables/disables authorization requirement for TBS"
```
As per the guidelines, please remove enables/disables:
https://docs.zephyrproject.org/latest/guides/build/kconfig/tips.html#prompt-strings

> For a Kconfig symbol that enables a driver/subsystem FOO, consider having just “Foo” as the prompt, instead of “Enable Foo support” or the like.

Maybe just write:
```suggestion
	bool "TBS authorization requirement"
```

tejlmand commented on subsys/bluetooth/audio/Kconfig.tbs:
```diff
@@ -152,4 +152,63 @@ config BT_DEBUG_TBS
+	  Call Control client.
+
+config BT_TBS_CLIENT_MAX_CALLS
+	int "Telephone Bearer Service client Client Maximum Number Of Calls Supported"
```
We are already in `BT_TBS_CLIENT` and the following settings doesn't include `Telephone Bearer Service client` in their titles, so maybe just:
```suggestion
	int "Maximum Number Of Calls Supported"
```

tejlmand commented on subsys/bluetooth/audio/Kconfig.tbs:
```diff
@@ -152,4 +152,63 @@ config BT_DEBUG_TBS
+
+##################### Call Control Client #####################
+
+config BT_TBS_CLIENT
```
Should this be a menuconfig instead, in order to give a better grouping in `menuconfig` with a subpage for all TBS_CLIENT related settings

```suggestion
menuconfig BT_TBS_CLIENT
```

Thalley replied:
For now I'd like to keep it as is, as that will follow the way the other LE Audio Kconfig files are defined. 

If we want to use `menuconfig` for any of them, we should do it for all of them as they follow the same pattern.

tejlmand commented on subsys/bluetooth/audio/Kconfig.tbs:
```diff
@@ -0,0 +1,155 @@
+
+#####################  Telephone Bearer Service #####################
+
+config BT_TBS
```
Should this be a menuconfig instead, in order to give a better grouping in `menuconfig` with a subpage for all BT_TBS related settings

```suggestion
menuconfig BT_TBS
```

Thalley commented:
Rebase to solve conflicts

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_audio/src/tbs_client_test.c:
```diff
@@ -0,0 +1,232 @@
+	discovery_complete = true;
+}
+
+static struct bt_tbs_client_cb tbs_client_cbs = {
```
Agreed. @asbjornsabo @Casper-Bonde-Bose @MariuszSkamra  Do you think we'll ever need to have more than 1 user registering LE Audio callbacks? In other words, should we use a list of callbacks, or a singleton?

asbjornsabo replied:
LE Audio callbacks in general, or TBS callbacks?
I believe a single user will be by far the most common case.  I suggest we stay with that until we see a need for anything else.

(Changing to a list will also mean significant changes elsewhere, which, if we want to do them, should be done after we have got the code int.)

Thalley replied:
I agree. 

If we change it for TBS, it would make sense to change it across the LE Audio stack. 

regardless, I'll go ahead and add `const` to the callbacks as suggested by @jhedberg

Thalley replied:
Changing it to `const` would require an API change here, and if we go with the `list` route `const` will need to be removed again. 

@jhedberg Can we keep as is for now?

Thalley replied:
Created https://github.com/zephyrproject-rtos/zephyr/issues/43913 to revisit this in a more general manner

jhedberg replied:
> Changing it to const would require an API change here, and if we go with the list route const will need to be removed again.

No, I don't think we'd take that path. Currently there are (if I counted correctly) 22 callback pointers in this struct. That's a non-trivial amount of memory. So if you want to add support for multiple instances in the future you'd either do that by an additional level of indirection (a top-level struct containing the linked list pointer and then a pointer to the const callbacks struct) or with the help of a dedicated linker section placed in ROM.

I don't see what's the problem here - adding const in a couple of places in this PR is a trivial thing to do.

Thalley replied:
> No, I don't think we'd take that path. Currently there are (if I counted correctly) 22 callback pointers in this struct. That's a non-trivial amount of memory. 

Correct, currently it is large. That can be significantly improved once https://github.com/zephyrproject-rtos/zephyr/issues/41216 is solved. 

> So if you want to add support for multiple instances in the future you'd either do that by an additional level of indirection (a top-level struct containing the linked list pointer and then a pointer to the const callbacks struct)

I don't see how that approach will provide any better memory. You'll still need a pointer for each callback function. 

> with the help of a dedicated linker section placed in ROM.

I am definitely fine with using ROM for the callbacks as well. 

Possibly we also want to move this into the `bt_tbs` struct, so that there is a callback struct for each TBS instance. 

> I don't see what's the problem here - adding const in a couple of places in this PR is a trivial thing to do.

I'll add them in this PR, and then when we revisit the callbacks we can continue the discussion there.

jhedberg replied:
> I don't see how that approach will provide any better memory. You'll still need a pointer for each callback function.

The struct with the callbacks would be placed in ROM, while the “parent” struct (with two pointers) would reside in RAM

Thalley committed:
Bluetooth: Audio: Make bt_tbs_client_cb const

Add `const` for the bt_tbs_client_cb as it is a singleton
that should not change over time.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

jhedberg commented on tests/bluetooth/bsim_bt/bsim_test_audio/src/tbs_client_test.c:
```diff
@@ -0,0 +1,232 @@
+		return;
+	}
+	printk("Connected to %s\n", addr);
+	tbs_conn = conn;
```
Missing bt_conn_ref() here?

Thalley replied:
Thanks for reminding me of applying the same fixes for these BSIM tests as I did for the others :) Remove the `tbs_conn` and updated to use the `default_conn` like the other LE Audio BSIM tests, as well as fixing the `bt_conn_ref`s.

jhedberg commented on tests/bluetooth/bsim_bt/bsim_test_audio/src/tbs_test.c:
```diff
@@ -0,0 +1,145 @@
+	}
+
+	printk("Connected to %s\n", addr);
+
```
Missing bt_conn_ref()?

jhedberg commented on tests/bluetooth/bsim_bt/bsim_test_audio/src/tbs_test.c:
```diff
@@ -0,0 +1,145 @@
+
+	printk("Connected to %s\n", addr);
+
+	tbs_conn = conn;
```
Missing bt_conn_ref()?

MariuszSkamra commented on subsys/bluetooth/audio/Kconfig.tbs:
```diff
@@ -0,0 +1,212 @@
+	  e.g. "tel" or "skype".
+	  Multiple values shall be comma (,) separated, e.g. "tel,skype".
+
+config BT_TBS_SIGNAL_STRENGTH_INTERVAL
+	int "Telephone Bearer Service Signal Strength Reporting Interval"
+	default 0
+	range 0 255
+	help
+	  Sets the interval of reporting the signal strength in seconds.
+		If the value is 0, the signal will not be reported.
```
As far as I remember the signal strength interval is configurable by the client, so why is there a need to have it in Kconfig?

Thalley replied:
This will be the initial value. Once we start registering the TBS dynamically this will be removed.

MariuszSkamra commented on subsys/bluetooth/audio/Kconfig.tbs:
```diff
@@ -0,0 +1,212 @@
+	  Sets the interval of reporting the signal strength in seconds.
+		If the value is 0, the signal will not be reported.
+
+config BT_TBS_STATUS_FLAGS
+	int "Telephone Bearer Service Features and Status value"
+	default 0
+	range 0 3
+	help
+	  Bitfield to set feature and status flags.
+	  	Bit 0: In-band ringtone
+		Bit 1: Silent mode
+		Bits 2-15: Reserved for future use
```
I have a feeling that flags should be maintained in runtime, not by Kconfig, e.g. silent mode could be enabled/disabled by button action

Thalley replied:
This will be the initial value. Once we start registering the TBS dynamically this will be removed.

MariuszSkamra commented on subsys/bluetooth/audio/Kconfig.tbs:
```diff
@@ -0,0 +1,212 @@
+		Bit 1: Silent mode
+		Bits 2-15: Reserved for future use
+
+config BT_TBS_SUPPORTED_FEATURES
+	int "Telephone Bearer Service Supported Features"
+	default 1
+	range 0 3
+	help
+	  Bitfield to set supported features of the bearer.
+		Bit 0: Local Hold and Retrieve
+		Bit 1: Join calls within Telephone Bearer Service
```
Shouldn't this be split to separate Kconfig settings?

Thalley replied:
Can you elaborate?

jhedberg replied:
@Thalley I guess @MariuszSkamra means that each feature bit would be its own boolean Kconfig option (which I agree would make this nicer for the user)

Thalley replied:
Ah, I see. Yeah that makes sense. 

@MariuszSkamra Is that OK to postpone for the dynamic registration task? I'd rather just get this merged as is so we can start on the other TBS/CCP related tasks as well

MariuszSkamra replied:
Sure

Thalley commented:
Rebased to solve conflicts
